### PR TITLE
scoped_truth for the loop variable being always less than the loop extent.

### LINKED ISSUE #8303: Simplifier fails to optimize for-loop endpoint.
I have no clue what's going on.

I'm seeing this:
image

Look at the `if (t84)` and trace the `t84` backwards:

 - `t84 = loop_var < t78`
 - `loop_var` goes from `0` to `t78`

So t84 should _always_ be false. It's the first time I'm seeing something like this.

I went back randomly to d9668c5bc to test and compare, but it produces the same result. I don't have time to make a repro right now, but wanted to document this already.

I'm not 100% sure, but I believe LLVM does notice this, and just removes the entire else-branch of this code. (It produces a non-vectorized else-branch).

### PR DESCRIPTION
**Pull request by mcourteaux**
Fixes #8303

### CODE REVIEW

**Comment by abadams** [comment]
This is actually a bit more subtle, because it's loop_var < min + extent, not loop_var < extent, but a scoped truth comparison on non-constants will really just look for that exact Expr. Maybe  `loop_var < mutate(min + extent, nullptr)`?

**Comment by mcourteaux** [comment]
Aren't loops rebased to have a zero starting range at that point?

**Comment by abadams** [comment]
The simplifier can run on almost any piece of IR, before or after loops get rebased to zero, so if you wanted to go with that strategy you'd have to add an `if (is_const_zero(op->min))`

**Comment by mcourteaux** [comment]
@abadams Okay, I think I fixed it. Thanks for helping out!

File: src/Simplify_Stmts.cpp [hunk 1]
```diff
@@ -240,7 +240,13 @@ Stmt Simplify::visit(const For *op) {
                                         loop_var_info);
 
         // If we're in the loop, the extent must be greater than 0.
-        ScopedFact fact = scoped_truth(extent_positive);
+        ScopedFact fact_extent_positive = scoped_truth(extent_positive);
+
+        // The loop variable will never exceed the loop bound.
+        Expr loop_var = Variable::make(Int(32), op->name);
+        Expr new_max = mutate(new_min + new_extent, nullptr);
+        ScopedFact fact_loop_var_less_than_extent = scoped_truth(loop_var < new_max);
```
Comment (abadams) [inline src/Simplify_Stmts.cpp:248]: Might be worth adding scoped_truth(new_min <= loop_var) while you're here.

Comment (mcourteaux) [reply src/Simplify_Stmts.cpp:248]: Good point. My brain is clearly not thinking this through... I committed that as well now.

**Review by abadams — APPROVED** [review]

**Pull request merged**